### ISSUE TITLE
如何在qlib里实现MACD金叉 条件判断呢？

### ISSUE THREAD

**Issue by DanielKui** [microsoft/qlib#2061]
我想实现MACD金叉因子，分析它的IC， 目前可以定义MACD , DIF, DEA的表达式，但如何增加金叉判断呢?

`if __name__ == "__main__":
    # 初始化qlib 数据源，默认为日线数据
    provider_uri = "~/.qlib/qlib_data/cn_data"  # target_dir
    qlib.init(provider_uri=provider_uri, region=REG_CN)

    #macd计算公式，分别计算DIFF DEA, MACD
    MACD_DIF = 'EMA($close, 12) - EMA($close, 26)'
    MACD_DEA = 'EMA((EMA($close, 12) - EMA($close, 26))/$close, 9)'
    MACD_EXP = '(EMA($close, 12) - EMA($close, 26))/$close - EMA((EMA($close, 12) - EMA($close, 26))/$close, 9)/$close'

    fields = [MACD_EXP, MACD_DIF, MACD_DEA,'$close']
    names = ['MACD', 'DIF', 'DEA', 'close']

    labels = ['Ref($close, -2)/Ref($close, -1) - 1'] # label
    label_names = ['LABEL']

    data_loader_config = {
        "feature": (fields, names),
        "label": (labels, label_names)
    }

    #SH600519可以是任何股票，但qlib目前官方的数据源不支持ETF，需要用我们clickhouse数据库的数据做转换,参考2.2.5 Qlib自定义数据源
    data_loader = QlibDataLoader(config=data_loader_config)
    df = data_loader.load(instruments=["SH600519"], start_time='2000-01-01', end_time='2020-08-01')

    #金叉
    #当快线（DIF）上穿慢线（DEA），并且整个结构在零轴之上时，往往是上涨趋势的确认，短线多头活跃
    cross_up =  (df['feature']['DIF'] > df['feature']['DEA'] ) & (df['feature']['DIF'].shift(1) <= df['feature']['DEA'].shift(1))
    cross_up.columns =  ['dt','instrument','macdresult']

    tmpUp = cross_up.to_frame()
    #print(tmpUp)#打印所有macresult,大部分为FALSE
    print("\r\n金叉:")
    
    print(tmpUp[tmpUp.iloc[:, -1]])#打印macresult为TRUE的行
`

如何放到DataHandlerLP 和 DatasetH里呢？然后进行IC统计

**Comment by SunsetWolf** [comment]
Hi, @DanielKui , Welcome to another question related to MACD.

- Converting the pandas golden-cross logic into a Qlib expression

  Your pandas version:
  ```
  cross_up = (df['feature']['DIF'] > df['feature']['DEA'] ) & (df['feature']['DIF'].shift(1) <= df['feature']['DEA'].shift(1))
  ```
  should be written in `qlib` as: 
  ```
  "(DIF > DEA) & (Ref(DIF, 1) <= Ref(DEA, 1))" 
  ```
  Note: where `Ref(x, 1)` is the forward shift.

- Add the golden cross factor to `data_handler_config` and use `DataHandlerLP` and `DatasetH` to handle loading the dataset

  Loading `DataHandlerLP` and `DatasetH` can be achieved by replacing `DIF` and `DEA` in the expression and adding them to `data_handler_config["data_loader"]["kwargs"]["config"]["feature"]`.

- Getting all golden-cross rows

  Once the dataset is prepared, identifying golden-cross points becomes straightforward:

  ```
  df[df[“gold”] > 0]
  ```

- Computing IC for the golden-cross factor

  You can compute the IC exactly the same way as your MACD IC:

  ```
  ic_series = dfmacd.groupby("datetime").apply(lambda x: x["MACD"].corr(x["LABEL"]))
  print("IC 指标统计：")
  print(f"IC 均值 = {ic_series.mean()}")
  print("IC 标准差：", ic_series.std())
  print("IC 绝对值均值：", ic_series.abs().mean())
  print(f"IC series: \n\n{ic_series}")
  ```

**Comment by DanielKui** [comment]
> Hi, [@DanielKui](https://github.com/DanielKui) , Welcome to another question related to MACD.
> 
> * Converting the pandas golden-cross logic into a Qlib expression
>   Your pandas version:
>   ```
>   cross_up = (df['feature']['DIF'] > df['feature']['DEA'] ) & (df['feature']['DIF'].shift(1) <= df['feature']['DEA'].shift(1))
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   should be written in `qlib` as:
>   ```
>   "(DIF > DEA) & (Ref(DIF, 1) <= Ref(DEA, 1))" 
>   ```
>   
>   
>       
>     
>   Note: where `Ref(x, 1)` is the forward shift.
> * Add the golden cross factor to `data_handler_config` and use `DataHandlerLP` and `DatasetH` to handle loading the dataset
>   Loading `DataHandlerLP` and `DatasetH` can be achieved by replacing `DIF` and `DEA` in the expression and adding them to `data_handler_config["data_loader"]["kwargs"]["config"]["feature"]`.
> * Getting all golden-cross rows
>   Once the dataset is prepared, identifying golden-cross points becomes straightforward:
>   ```
>   df[df[“gold”] > 0]
>   ```
> * Computing IC for the golden-cross factor
>   You can compute the IC exactly the same way as your MACD IC:
>   ```
>   ic_series = dfmacd.groupby("datetime").apply(lambda x: x["MACD"].corr(x["LABEL"]))
>   print("IC 指标统计：")
>   print(f"IC 均值 = {ic_series.mean()}")
>   print("IC 标准差：", ic_series.std())
>   print("IC 绝对值均值：", ic_series.abs().mean())
>   print(f"IC series: \n\n{ic_series}")
>   ```

Thank you very much for you detailed answer.


fields = [MACD_EXP, MACD_DIF, MACD_DEA,'$close','($DIF > $DEA) & (Ref($DIF, 1) <= Ref($DEA, 1))']
names = ['MACD', 'DIF', 'DEA', 'close','gold']

    data_handler_config = {
        "start_time": "2008-01-01",
        "end_time": "2020-08-01",
        "instruments": "all",
        "data_loader": {
            "class": "QlibDataLoader",  # Assuming QlibDataLoader is a string reference to the class
            "kwargs": {
                "config": {
                    "feature": (fields, names),
                    "label": (labels, label_names),
                },
                "freq": "day",
            },
        },
    }

是这么定义金叉fields么？传递到data_handler_config["data_loader"]["kwargs"]["config"]["feature"]？
我这么做貌似失败了
<img width="1965" height="403" alt="Image" src="https://github.com/user-attachments/assets/9c86e7bb-e14e-48b2-84a5-58de798c63dc" />

**Comment by SunsetWolf** [comment]
Hi, @DanielKui , I think you have to replace `DIF` and `DEA` here in `"(DIF > DEA) & (Ref(DIF, 1) <= Ref(DEA, 1))" ` with the appropriate expressions to make it work.

**Comment by DanielKui** [comment]
> Hi, [@DanielKui](https://github.com/DanielKui) , I think you have to replace `DIF` and `DEA` here in `"(DIF > DEA) & (Ref(DIF, 1) <= Ref(DEA, 1))" ` with the appropriate expressions to make it work.

Thank you very much.

再问一个问题不好意思啊
如果想针对某一个特定的股票进行金叉的IC分析该如何实现呢？
`
if __name__ == "__main__":
    # 初始化qlib 数据源，默认为日线数据
    provider_uri = "~/.qlib/qlib_data/cn_data"  # target_dir
    qlib.init(provider_uri=provider_uri, region=REG_CN)

    #macd计算公式，分别计算DIFF DEA, MACD
    MACD_DIF = 'EMA($close, 12) - EMA($close, 26)'
    MACD_DEA = 'EMA((EMA($close, 12) - EMA($close, 26))/$close, 9)'
    MACD_EXP = '(EMA($close, 12) - EMA($close, 26))/$close - EMA((EMA($close, 12) - EMA($close, 26))/$close, 9)/$close'

    fields = [MACD_EXP, MACD_DIF, MACD_DEA,'$close']
    names = ['MACD', 'DIF', 'DEA', 'close']

    labels = ['Ref($close, -2)/Ref($close, -1) - 1'] # label
    label_names = ['LABEL']

    data_loader_config = {
        "feature": (fields, names),
        "label": (labels, label_names)
    }

    #SH600519可以是任何股票，但qlib目前官方的数据源不支持ETF，需要用我们clickhouse数据库的数据做转换,参考2.2.5 Qlib自定义数据源
    data_loader = QlibDataLoader(config=data_loader_config)
    df = data_loader.load(instruments=["SH600519"], start_time='2000-01-01', end_time='2020-08-01')

    #金叉
    #当快线（DIF）上穿慢线（DEA），并且整个结构在零轴之上时，往往是上涨趋势的确认，短线多头活跃
    cross_up =  df[(df['feature']['DIF'] > df['feature']['DEA'] ) & (df['feature']['DIF'].shift(1) <= df['feature']['DEA'].shift(1))]
    print(cross_up)

    ic_series = cross_up.groupby("datetime").apply(lambda x: x['feature']['MACD'].corr(x['label']['LABEL']))
    
    print("金叉IC 指标统计：")
    print(f"IC 均值 = {ic_series.mean()}")
    print("IC 标准差：", ic_series.std())
    print("IC 绝对值均值：", ic_series.abs().mean())
    print(f"IC series: \n\n{ic_series}")

    print("\r\n\r\n=====================\r\n\r\n")`


<img width="1662" height="855" alt="Image" src="https://github.com/user-attachments/assets/8f9e6bbd-f0f5-4453-8852-feb4bb80ea25" />

我这么实现的，但是IC统计结果是0，哪个环节出错了呢？ 
也是使用的这个语句进行的IC统计
ic_series = cross_up.groupby("datetime").apply(lambda x: x['feature']['MACD'].corr(x['label']['LABEL']))

 x['feature']['MACD']和x['label']['LABEL']都是有值的，为何会divided by 0呢？

@SunsetWolf

**Comment by SunsetWolf** [comment]
Hi, @DanielKui , `IC` is the “cross-sectional correlation” 
which is the correlation between the “factor of multiple stocks on the same day vs. the future returns of multiple stocks”. ". But now you only have one stock. So a statistic of 0 is normal. This is not an error.

**Comment by DanielKui** [comment]
> Hi, [@DanielKui](https://github.com/DanielKui) , `IC` is the “cross-sectional correlation” which is the correlation between the “factor of multiple stocks on the same day vs. the future returns of multiple stocks”. ". But now you only have one stock. So a statistic of 0 is normal. This is not an error.

Thank you very much.
